File: projects/engineering.purdue.edu/kak/programmingwithobjects/pwocode/chap3/NestedClass.cc
// This code example is from the following source:
//
// Book Title:  Programming with Objects, A Comparative Presentation
//              of Object Oriented Programming with C++ and Java
//
// Book Author:  Avinash Kak
//
// Chapter:     Chapter 3 ---- The Notion Of A Class And Some Other Ideas
//
// Section:     Section 3.16 - Nested Types
//





//NestedClass.cc

#include <iostream>
using namespace std;

class X {                                                         //(A)

    class Y{                                                      //(B)
        int m;
    public:
        Y( int mm ) { m = mm; }                                   //(C)
        void printY(){ cout << "m of nested class object: " 
                            << m << endl; };                      //(D)
    };

    Y* yptr;                                                      //(E)
public:
    X() { yptr = new Y( 100 ); }                                  //(F)
    Y* get_yptr(){ return yptr; }                                 //(G)
};


int main() {                           
    X x;                                                          //(H)
    x.get_yptr()->printY();                                       //(I)
    return 0;
}

File: projects/engineering.purdue.edu/programmingwithobjects/pwocode/chap3/NestedClass.java
// This code example is from the following source:
//
// Book Title:  Programming with Objects, A Comparative Presentation
//              of Object Oriented Programming with C++ and Java
//
// Book Author:  Avinash Kak
//
// Chapter:     Chapter 3 ---- The Notion Of A Class And Some Other Key Ideas
//
// Section:     Section 3.16 - Nested Types
//




//NestedClass.java

class X {                                                         //(A)
    static class Y{                                               //(B)
        private int m;
        public Y( int mm ) { m = mm; }      
        public void printY(){               
            System.out.println( "m of nested class object: " + m );
        }
    }

    private Y yref;                         

    public X() { yref = new Y( 100 ); }     

    Y get_yref(){ return yref; }            
}

class Test {
    public static void main( String[] args ) {
        X x = new X();
        x.get_yref().printY();   // m of nested class object: 100 
    }
}
